compiler/src/iree/compiler/Codegen/Common/TileAndDecomposeWinogradInputPass.cpp

```diff
@@ -0,0 +1,256 @@
+
+  static void computeLoopParams(SmallVectorImpl<Value> &lbs,
+                                SmallVectorImpl<Value> &ubs,
+                                SmallVectorImpl<Value> &steps,
+                                ArrayRef<int64_t> tensorShape, Location loc,
+                                PatternRewriter &rewriter) {
+    auto zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    for (auto shape : llvm::enumerate(tensorShape)) {
+      if (shape.index() < 2) continue;
+      lbs.push_back(zero);
+      ubs.push_back(
+          rewriter.create<arith::ConstantIndexOp>(loc, shape.value()));
+      steps.push_back(one);
+    }
+  }
+
+  LogicalResult matchAndRewrite(
+      IREE::LinalgExt::WinogradInputTransformOp inputOp,
+      PatternRewriter &rewriter) const override {
+    auto loc = inputOp.getLoc();
+    auto funcOp = inputOp->getParentOfType<func::FuncOp>();
+    if (!funcOp) return failure();
+
+    const float *BT{nullptr};
+    auto inputTileSize = inputOp.getInputTileSize();
+    auto outputTileSize = inputOp.getOutputTileSize();
+    switch (outputTileSize) {
+      case 6:
+        BT = IREE::LinalgExt::Winograd::BT_6x6_3x3;
+        break;
+      default:
+        return failure();
+    }
+    Value BTV = IREE::LinalgExt::createValueFrom2DConstant(
+        BT, inputTileSize, inputTileSize, false, loc, rewriter);
+    Value BV = IREE::LinalgExt::createValueFrom2DConstant(
+        BT, inputTileSize, inputTileSize, true, loc, rewriter);
+
+    auto input = inputOp.input();
+    auto output = inputOp.output();
+    auto outputType = output.getType().cast<ShapedType>();
+    auto inputType = input.getType().cast<ShapedType>();
+    auto outputShape = outputType.getShape();
+    auto inputShape = inputType.getShape();
+    auto elementType = outputType.getElementType();
+    SmallVector<int64_t> inputTileSquare(2, inputTileSize);
+
+    rewriter.setInsertionPointToStart(&funcOp.getBody().front());
+    auto zeroF32 = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(elementType));
+    auto scratch =
+        rewriter.create<tensor::EmptyOp>(loc, inputTileSquare, elementType);
+
+    rewriter.setInsertionPoint(inputOp);
+    SmallVector<Value> lbs, ubs, steps;
+    computeLoopParams(lbs, ubs, steps, outputShape, loc, rewriter);
+    // Construct loops
+    auto loopNest = scf::buildLoopNest(
+        rewriter, loc, lbs, ubs, steps, ValueRange({output}),
+        [&](OpBuilder &nestedBuilder, Location loc, ValueRange outputIvs,
+            ValueRange iterArgs) -> scf::ValueVector { return {iterArgs[0]}; });
+
+    // Add spir-v attributes to loops (H, W, C)
+    const char *attrName = "iree.spirv.distribute_dim";
+    for (int i = loopNest.loops.size() - 1, dim = 0; i > 0; --i) {
+      loopNest.loops[i]->setAttr(attrName, rewriter.getIndexAttr(dim++));
+    }
+
+    // Extract input slice
+    auto one = rewriter.getIndexAttr(1);
```

Comment:
A lot of the code here is effectively the same as the loop generation with tiling.... Thinking out loud here, but maybe if we treat this as "double tiling", then we can implement the first level tiling as is today, and use an additional level of tiling to implement this.... Just a thought...

Reply:
Totally agree - this is essentially just loop generating with tiling. The only differences I can see are that for the H and W dims the affine maps are different from what is typically generated and the extract slice depends on these affine maps (which means we can't use the existing extract slice mechanism?). Also, the other important factor is that I have only exposed the batch dim (N) and channel dim (C) to the tiling interface.  So you are right I could do the second level of tiling on N and C using the existing tiling infra, but wouldn't this introduce an additional extract_slice between the second-level tiling C loop and the H and W loops? Unless there is a way to make it go away?